File: plugins/plugin-linear/python/src/elizaos_plugin_linear/actions/list_teams.py
import json
import logging
import re
from typing import Any

from elizaos_plugin_linear.actions.base import (
    ActionExample,
    ActionResult,
    HandlerCallback,
    Memory,
    RuntimeProtocol,
    State,
    create_action,
)
from elizaos_plugin_linear.services.linear import LinearService

logger = logging.getLogger(__name__)

LIST_TEAMS_TEMPLATE = """Extract team filter criteria from the user's request.

User request: "{user_message}"

Return ONLY a JSON object:
{{
  "nameFilter": "Keywords to search in team names",
  "specificTeam": "Specific team name or key if looking for one team",
  "showAll": true/false (true if user explicitly asks for "all"),
  "includeDetails": true/false (true if user wants detailed info)
}}

Only include fields that are clearly mentioned."""


async def validate(
    runtime: RuntimeProtocol,
    _message: Memory,
    _state: State | None = None,
) -> bool:
    try:
        api_key = runtime.get_setting("LINEAR_API_KEY")
        return bool(api_key)
    except Exception:
        return False


async def handler(
    runtime: RuntimeProtocol,
    message: Memory,
    _state: State | None = None,
    options: dict[str, Any] | None = None,
    callback: HandlerCallback | None = None,
) -> ActionResult:
    try:
        linear_service: LinearService = runtime.get_service("linear")
        if not linear_service:
            raise RuntimeError("Linear service not available")

        content = message.get("content", {}).get("text", "")
        name_filter: str | None = None
        specific_team: str | None = None

        if content:
            prompt = LIST_TEAMS_TEMPLATE.format(user_message=content)
            response = await runtime.use_model("TEXT_LARGE", {"prompt": prompt})

            if response:
                try:
                    cleaned = re.sub(r"^```(?:json)?\n?", "", response)
                    cleaned = re.sub(r"\n?```$", "", cleaned).strip()
                    parsed = json.loads(cleaned)

                    name_filter = parsed.get("nameFilter")
                    specific_team = parsed.get("specificTeam")
                    parsed.get("includeDetails", False)

                except json.JSONDecodeError:
                    logger.warning("Failed to parse team filters")

        teams = await linear_service.get_teams()

        # Filter for specific team
        if specific_team:
            teams = [
                t
                for t in teams
                if t["key"].lower() == specific_team.lower()
                or t["name"].lower() == specific_team.lower()
            ]

        # Filter by name keywords
        if name_filter and not specific_team:
            keywords = name_filter.lower().split()
            teams = [
                t
                for t in teams
                if any(kw in f"{t['name']} {t.get('description', '')}".lower() for kw in keywords)
            ]

        if not teams:
            no_teams_msg = (
                f'No team found matching "{specific_team}".'
                if specific_team
                else f'No teams found matching "{name_filter}".'
                if name_filter
                else "No teams found in Linear."
            )
            if callback:
                await callback(
                    {"text": no_teams_msg, "source": message.get("content", {}).get("source")}
                )
            return {"text": no_teams_msg, "success": True, "data": {"teams": []}}

        team_list = []
        for i, team in enumerate(teams):
            info = f"{i + 1}. {team['name']} ({team['key']})"

            if team.get("description"):
                info += f"\n   {team['description']}"

            team_list.append(info)

        header_text = (
            "📋 Team Details:"
            if specific_team and len(teams) == 1
            else f'📋 Found {len(teams)} team{"s" if len(teams) != 1 else ""} matching "{name_filter}":'
            if name_filter
            else f"📋 Found {len(teams)} team{'s' if len(teams) != 1 else ''}:"
        )

        result_message = f"{header_text}\n\n" + "\n\n".join(team_list)
        if callback:
            await callback(
                {"text": result_message, "source": message.get("content", {}).get("source")}
            )

        return {
            "text": f"Found {len(teams)} team{'s' if len(teams) != 1 else ''}",
            "success": True,
            "data": {
                "teams": [
                    {
                        "id": t["id"],
                        "name": t["name"],
                        "key": t["key"],
                        "description": t.get("description"),
                    }
                    for t in teams
                ],
                "count": len(teams),
                "filters": {"name": name_filter, "specific": specific_team},
            },
        }

    except Exception as error:
        logger.error(f"Failed to list teams: {error}")
        error_message = f"❌ Failed to list teams: {error}"
        if callback:
            await callback(
                {"text": error_message, "source": message.get("content", {}).get("source")}
            )
        return {"text": error_message, "success": False}


list_teams_action = create_action(
    name="LIST_LINEAR_TEAMS",
    description="List teams in Linear with optional filters",
    similes=["list-linear-teams", "show-linear-teams", "get-linear-teams"],
    examples=[
        [
            ActionExample(name="User", content={"text": "Show me all teams"}),
            ActionExample(
                name="Assistant",
                content={
                    "text": "I'll list all the teams in Linear.",
                    "actions": ["LIST_LINEAR_TEAMS"],
                },
            ),
        ],
    ],
    validate=validate,
    handler=handler,
)

File: plugins/plugin-linear/src/actions/listTeams.ts
import {
  type Action,
  type ActionResult,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  logger,
  type Memory,
  ModelType,
  type State,
} from "@elizaos/core";
import type { Team, User } from "@linear/sdk";
import { listTeamsTemplate } from "../generated/prompts/typescript/prompts.js";
import type { LinearService } from "../services/linear";

interface TeamWithDetails extends Team {
  memberCount?: number;
  projectCount?: number;
  membersList?: User[];
}

export const listTeamsAction: Action = {
  name: "LIST_LINEAR_TEAMS",
  description: "List teams in Linear with optional filters",
  similes: ["list-linear-teams", "show-linear-teams", "get-linear-teams", "view-linear-teams"],

  examples: [
    [
      {
        name: "User",
        content: {
          text: "Show me all teams",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll list all the teams in Linear for you.",
          actions: ["LIST_LINEAR_TEAMS"],
        },
      },
    ],
    [
      {
        name: "User",
        content: {
          text: "Which engineering teams do we have?",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "Let me find the engineering teams for you.",
          actions: ["LIST_LINEAR_TEAMS"],
        },
      },
    ],
    [
      {
        name: "User",
        content: {
          text: "Show me the teams I'm part of",
        },
      },
      {
        name: "Assistant",
        content: {
          text: "I'll show you the teams you're a member of.",
          actions: ["LIST_LINEAR_TEAMS"],
        },
      },
    ],
  ],

  validate: async (runtime: any, message: any, state?: any, options?: any): Promise<boolean> => {
    const __avTextRaw = typeof message?.content?.text === "string" ? message.content.text : "";
    const __avText = __avTextRaw.toLowerCase();
    const __avKeywords = ["list", "linear", "teams"];
    const __avKeywordOk =
      __avKeywords.length > 0 &&
      __avKeywords.some((word) => word.length > 0 && __avText.includes(word));
    const __avRegex = /\b(?:list|linear|teams)\b/i;
    const __avRegexOk = __avRegex.test(__avText);
    const __avSource = String(message?.content?.source ?? message?.source ?? "");
    const __avExpectedSource = "";
    const __avSourceOk = __avExpectedSource
      ? __avSource === __avExpectedSource
      : Boolean(
          __avSource || state || runtime?.agentId || runtime?.getService || runtime?.getSetting
        );
    const __avOptions = options && typeof options === "object" ? options : {};
    const __avInputOk =
      __avText.trim().length > 0 ||
      Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
      Boolean(message?.content && typeof message.content === "object");

    if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
      return false;
    }

    const __avLegacyValidate = async (
      runtime: any,
      message: any,
      state?: any,
      options?: any
    ): Promise<boolean> => {
      const __avTextRaw = typeof message?.content?.text === "string" ? message.content.text : "";
      const __avText = __avTextRaw.toLowerCase();
      const __avKeywords = ["list", "linear", "teams"];
      const __avKeywordOk =
        __avKeywords.length > 0 &&
        __avKeywords.some((word) => word.length > 0 && __avText.includes(word));
      const __avRegex = /\b(?:list|linear|teams)\b/i;
      const __avRegexOk = __avRegex.test(__avText);
      const __avSource = String(message?.content?.source ?? message?.source ?? "");
      const __avExpectedSource = "";
      const __avSourceOk = __avExpectedSource
        ? __avSource === __avExpectedSource
        : Boolean(
            __avSource || state || runtime?.agentId || runtime?.getService || runtime?.getSetting
          );
      const __avOptions = options && typeof options === "object" ? options : {};
      const __avInputOk =
        __avText.trim().length > 0 ||
        Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
        Boolean(message?.content && typeof message.content === "object");

      if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
        return false;
      }

      const __avLegacyValidate = async (
        runtime: any,
        message: any,
        state?: any,
        options?: any
      ): Promise<boolean> => {
        const __avTextRaw = typeof message?.content?.text === "string" ? message.content.text : "";
        const __avText = __avTextRaw.toLowerCase();
        const __avKeywords = ["list", "linear", "teams"];
        const __avKeywordOk =
          __avKeywords.length > 0 &&
          __avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
        const __avRegex = /\b(?:list|linear|teams)\b/i;
        const __avRegexOk = __avRegex.test(__avText);
        const __avSource = String(message?.content?.source ?? message?.source ?? "");
        const __avExpectedSource = "";
        const __avSourceOk = __avExpectedSource
          ? __avSource === __avExpectedSource
          : Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
        const __avOptions = options && typeof options === "object" ? options : {};
        const __avInputOk =
          __avText.trim().length > 0 ||
          Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
          Boolean(message?.content && typeof message.content === "object");

        if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
          return false;
        }

        const __avLegacyValidate = async (
          runtime: IAgentRuntime,
          _message: Memory,
          _state?: State
        ) => {
          const apiKey = runtime.getSetting("LINEAR_API_KEY");
          return !!apiKey;
        };
        try {
          return Boolean(await (__avLegacyValidate as any)(runtime, message, state, options));
        } catch {
          return false;
        }
      };
      try {
        return Boolean(await (__avLegacyValidate as any)(runtime, message, state, options));
      } catch {
        return false;
      }
    };
    try {
      return Boolean(await (__avLegacyValidate as any)(runtime, message, state, options));
    } catch {
      return false;
    }
  },

  async handler(
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult> {
    try {
      const linearService = runtime.getService<LinearService>("linear");
      if (!linearService) {
        throw new Error("Linear service not available");
      }

      const content = message.content.text || "";
      let nameFilter: string | undefined;
      let specificTeam: string | undefined;
      let myTeams = false;
      let includeDetails = false;

      if (content) {
        const prompt = listTeamsTemplate.replace("{{userMessage}}", content);
        const response = await runtime.useModel(ModelType.TEXT_LARGE, {
          prompt: prompt,
        });

        if (response) {
          try {
            const parsed = JSON.parse(
              response
                .replace(/^```(?:json)?\n?/, "")
                .replace(/\n?```$/, "")
                .trim()
            );

            nameFilter = parsed.nameFilter;
            specificTeam = parsed.specificTeam;
            myTeams = parsed.myTeams === true;
            includeDetails = parsed.includeDetails === true;
          } catch (parseError) {
            logger.warn("Failed to parse team filters:", parseError);
          }
        }
      }

      let teams = await linearService.getTeams();

      if (specificTeam) {
        teams = teams.filter(
          (team) =>
            team.key.toLowerCase() === specificTeam.toLowerCase() ||
            team.name.toLowerCase() === specificTeam.toLowerCase()
        );
      }

      if (nameFilter && !specificTeam) {
        const keywords = nameFilter.toLowerCase().split(/\s+/);
        teams = teams.filter((team) => {
          const teamText = `${team.name} ${team.description || ""}`.toLowerCase();
          return keywords.some((keyword) => teamText.includes(keyword));
        });
      }

      if (myTeams) {
        try {
          const userTeams = await linearService.getUserTeams();
          const userTeamIds = new Set(userTeams.map((t) => t.id));
          teams = teams.filter((team) => userTeamIds.has(team.id));
        } catch (error) {
          logger.warn("Could not filter for user's teams:", error);
        }
      }

      if (teams.length === 0) {
        const noTeamsMessage = specificTeam
          ? `No team found matching "${specificTeam}".`
          : nameFilter
            ? `No teams found matching "${nameFilter}".`
            : "No teams found in Linear.";
        await callback?.({
          text: noTeamsMessage,
          source: message.content.source,
        });
        return {
          text: noTeamsMessage,
          success: true,
          data: {
            teams: [],
          },
        };
      }

      let teamsWithDetails: TeamWithDetails[] = teams;
      if (includeDetails || specificTeam) {
        teamsWithDetails = await Promise.all(
          teams.map(async (team): Promise<TeamWithDetails> => {
            const membersQuery = await team.members();
            const members = await membersQuery.nodes;
            const projectsQuery = await team.projects();
            const projects = await projectsQuery.nodes;

            return Object.assign(team, {
              memberCount: members.length,
              projectCount: projects.length,
              membersList: specificTeam ? members.slice(0, 5) : [],
            });
          })
        );
      }

      const teamList = teamsWithDetails
        .map((team: TeamWithDetails, index: number) => {
          let info = `${index + 1}. ${team.name} (${team.key})`;

          if (team.description) {
            info += `\n   ${team.description}`;
          }

          if (includeDetails || specificTeam) {
            info += `\n   Members: ${team.memberCount ?? 0} | Projects: ${team.projectCount ?? 0}`;

            const membersList = team.membersList ?? [];
            if (specificTeam && membersList.length > 0) {
              const memberNames = membersList.map((m: User) => m.name).join(", ");
              info += `\n   Team members: ${memberNames}${(team.memberCount ?? 0) > 5 ? " ..." : ""}`;
            }
          }

          return info;
        })
        .join("\n\n");

      const headerText =
        specificTeam && teams.length === 1
          ? `📋 Team Details:`
          : nameFilter
            ? `📋 Found ${teams.length} team${teams.length === 1 ? "" : "s"} matching "${nameFilter}":`
            : `📋 Found ${teams.length} team${teams.length === 1 ? "" : "s"}:`;

      const resultMessage = `${headerText}\n\n${teamList}`;
      await callback?.({
        text: resultMessage,
        source: message.content.source,
      });

      return {
        text: `Found ${teams.length} team${teams.length === 1 ? "" : "s"}`,
        success: true,
        data: {
          teams: teamsWithDetails.map((t: TeamWithDetails) => ({
            id: t.id,
            name: t.name,
            key: t.key,
            description: t.description,
            memberCount: t.memberCount,
            projectCount: t.projectCount,
          })),
          count: teams.length,
          filters: {
            name: nameFilter,
            specific: specificTeam,
          },
        },
      };
    } catch (error) {
      logger.error("Failed to list teams:", error);
      const errorMessage = `❌ Failed to list teams: ${error instanceof Error ? error.message : "Unknown error"}`;
      await callback?.({
        text: errorMessage,
        source: message.content.source,
      });
      return {
        text: errorMessage,
        success: false,
      };
    }
  },
};
